src/Symfony/Component/HttpFoundation/File/MimeType/MimeTypeGuesser.php

```diff
@@ -125,18 +117,40 @@ public function guess($path)
             throw new AccessDeniedException($path);
         }
 
-        if (!$this->guessers) {
+        if (!$guessers = $this->getGuessers()) {
             $msg = 'Unable to guess the mime type as no guessers are available';
             if (!FileinfoMimeTypeGuesser::isSupported()) {
                 $msg .= ' (Did you enable the php_fileinfo extension?)';
             }
             throw new \LogicException($msg);
         }
 
-        foreach ($this->guessers as $guesser) {
+        foreach ($guessers as $guesser) {
             if (null !== $mimeType = $guesser->guess($path)) {
                 return $mimeType;
             }
         }
     }
+
+    /**
+     * @return MimeTypeGuesserInterface[]
+     */
+    private function getGuessers()
+    {
+        if (!$this->nativeGuessersLoaded) {
+            // register all natively provided mime type guessers.
+
+            if (FileinfoMimeTypeGuesser::isSupported()) {
```

Comment:
Yes, I understand what you are proposing. Yes, this can be done once while compilation. But then it's better not to use a Singleton and have just a normal ChainGuesser, empty by default. The problem is that `MimeTypeGuesser` as a Singleton is already used everywhere, so it's easier to fix it this way.